File: lib/ruby/cosmos-llm-ruby-context/lib/cosmos/llm/context/block.rb
# frozen_string_literal: true

module Cosmos
  module Llm
    module Context
      # Represents a named block of content within an LLM context.
      #
      # Context blocks are flexible containers for different types of content that
      # make up an agentic context. They can represent system prompts, user messages,
      # tool declarations, arbitrary strings, or any other structured content.
      #
      # @example Create a system prompt block
      #   block = Block.new(:system, "You are a helpful assistant")
      #
      # @example Create a tool declaration block
      #   block = Block.new(:tool, { name: "calculator", parameters: [...] })
      #
      # @see Builder
      class Block
        # @return [Symbol, String] The name/type of this block
        attr_reader :name

        # @return [Object] The content of this block
        attr_reader :content

        # @return [Hash] Additional metadata for this block
        attr_reader :metadata

        # Initializes a new Block.
        #
        # @param name [Symbol, String] The name/type of the block
        # @param content [Object] The content of the block
        # @param metadata [Hash] Optional metadata for the block
        # @return [Block] A new block instance
        # @raise [InvalidNameError] If name is invalid
        # @raise [ValidationError] If metadata is not a Hash
        def initialize(name, content, metadata = {})
          @name = validate_and_coerce_name(name)
          @content = content
          @metadata = validate_metadata(metadata).freeze
          freeze
        end

        # Gets a metadata value.
        #
        # Since blocks are immutable, metadata cannot be modified after creation.
        # To create a block with different metadata, create a new Block instance.
        #
        # @param key [Symbol, String] The metadata key
        # @return [Object, nil] The metadata value, or nil if not found
        # @example Get metadata
        #   block.meta(:role) # => 'system'
        def meta(key)
          @metadata[key]
        end

        # Creates a new Block with updated metadata.
        #
        # @param updates [Hash] The metadata updates to apply
        # @return [Block] A new Block instance with merged metadata
        # @example Update metadata
        #   new_block = block.with_metadata(role: 'user')
        def with_metadata(updates)
          Block.new(@name, @content, @metadata.merge(updates))
        end

        # Creates a new Block with updated content.
        #
        # @param new_content [Object] The new content
        # @return [Block] A new Block instance with updated content
        # @example Update content
        #   new_block = block.with_content("new content")
        def with_content(new_content)
          Block.new(@name, new_content, @metadata)
        end

        # Checks if this is a specific type of block.
        #
        # @param block_type [Symbol, String] The type to check
        # @return [Boolean] True if the block matches the type
        # @example Check block type
        #   block = Block.new(:system, "...")
        #   block.type?(:system) # => true
        #   block.type?(:user) # => false
        def type?(block_type)
          @name.to_sym == block_type.to_sym
        end

        # Converts the block to a hash representation.
        #
        # @return [Hash] Hash with name, content, and metadata
        def to_h
          {
            name: @name,
            content: @content,
            metadata: @metadata
          }
        end

        # Returns a string representation of the block.
        #
        # @return [String] String representation
        def to_s
          content_preview = @content.to_s.length > 50 ? "#{@content.to_s[0..46]}..." : @content.to_s
          "#<Block:#{@name} content=\"#{content_preview}\">"
        end

        # Inspects the block.
        #
        # @return [String] Detailed string representation
        def inspect
          "#<Cosmos::Llm::Context::Block:0x#{object_id.to_s(16)} " \
            "@name=#{@name.inspect} @content=#{@content.inspect} " \
            "@metadata=#{@metadata.inspect}>"
        end

        # Implements equality comparison.
        #
        # @param other [Object] The object to compare with
        # @return [Boolean] True if the blocks are equal
        def ==(other)
          return false unless other.is_a?(Block)

          name == other.name &&
            content == other.content &&
            metadata == other.metadata
        end

        alias eql? ==

        # Generates a hash code for the block.
        #
        # @return [Integer] The hash code
        def hash
          [name, content, metadata].hash
        end

        private

        # Validates and coerces the name to a Symbol.
        #
        # @param name [Object] The name to validate
        # @return [Symbol] The validated name as a Symbol
        # @raise [InvalidNameError] If the name is invalid
        def validate_and_coerce_name(name)
          raise InvalidNameError, 'Name cannot be nil' if name.nil?

          case name
          when Symbol
            name
          when String
            raise InvalidNameError, 'Name cannot be empty' if name.empty?

            name.to_sym
          else
            raise InvalidNameError, "Name must be a String or Symbol, got #{name.class}"
          end
        end

        # Validates that metadata is a Hash.
        #
        # @param metadata [Object] The metadata to validate
        # @return [Hash] The validated metadata
        # @raise [ValidationError] If metadata is not a Hash
        def validate_metadata(metadata)
          raise ValidationError, "Metadata must be a Hash, got #{metadata.class}" unless metadata.is_a?(Hash)

          metadata
        end
      end
    end
  end
end

# Copyright (c) 2025 Durable Programming, LLC. All rights reserved.

File: lib/rust/cosmos-llm-rust-context/src/block.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::error::ContextError;

/// An immutable, named content block within an LLM context.
///
/// Blocks can represent system prompts, user messages, tool declarations,
/// or any other labelled content. Once constructed they are immutable;
/// use [`with_content`] and [`with_metadata`] to derive modified copies.
///
/// # Examples
///
/// ```rust
/// use std::collections::HashMap;
/// use cosmos_llm_context::Block;
///
/// let b = Block::new("system", "You are helpful.", HashMap::new()).unwrap();
/// assert_eq!(b.name, "system");
/// assert!(b.is_type("system"));
/// ```
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Block {
    /// The block type / role name (e.g. `"system"`, `"user"`).
    pub name: String,
    /// The block's text content.
    pub content: String,
    /// Arbitrary metadata attached to this block.
    pub metadata: HashMap<String, Value>,
}

impl Block {
    /// Creates a new block.
    ///
    /// # Errors
    ///
    /// Returns [`ContextError::InvalidName`] if `name` is empty.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use std::collections::HashMap;
    /// use cosmos_llm_context::Block;
    ///
    /// let b = Block::new("user", "Hello!", HashMap::new()).unwrap();
    /// assert_eq!(b.content, "Hello!");
    /// ```
    pub fn new(
        name: impl Into<String>,
        content: impl Into<String>,
        metadata: HashMap<String, Value>,
    ) -> Result<Self, ContextError> {
        let name = name.into();
        if name.is_empty() {
            return Err(ContextError::InvalidName("name cannot be empty".into()));
        }
        Ok(Self {
            name,
            content: content.into(),
            metadata,
        })
    }

    /// Returns `true` if this block's name matches `block_type` (case-sensitive).
    ///
    /// # Examples
    ///
    /// ```rust
    /// use std::collections::HashMap;
    /// use cosmos_llm_context::Block;
    ///
    /// let b = Block::new("system", "...", HashMap::new()).unwrap();
    /// assert!(b.is_type("system"));
    /// assert!(!b.is_type("user"));
    /// ```
    pub fn is_type(&self, block_type: &str) -> bool {
        self.name == block_type
    }

    /// Gets a metadata value.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use std::collections::HashMap;
    /// use cosmos_llm_context::Block;
    /// use serde_json::json;
    ///
    /// let meta = [("priority".to_string(), json!(1))].into();
    /// let b = Block::new("user", "hi", meta).unwrap();
    /// assert_eq!(b.meta("priority"), Some(&json!(1)));
    /// ```
    pub fn meta(&self, key: &str) -> Option<&Value> {
        self.metadata.get(key)
    }

    /// Returns a copy of this block with updated content.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use std::collections::HashMap;
    /// use cosmos_llm_context::Block;
    ///
    /// let b = Block::new("user", "old", HashMap::new()).unwrap();
    /// let b2 = b.with_content("new");
    /// assert_eq!(b2.content, "new");
    /// ```
    pub fn with_content(&self, content: impl Into<String>) -> Self {
        Self {
            name: self.name.clone(),
            content: content.into(),
            metadata: self.metadata.clone(),
        }
    }

    /// Returns a copy of this block with merged metadata.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use std::collections::HashMap;
    /// use cosmos_llm_context::Block;
    /// use serde_json::json;
    ///
    /// let b = Block::new("system", "hi", HashMap::new()).unwrap();
    /// let b2 = b.with_metadata([("role".to_string(), json!("system"))].into());
    /// assert!(b2.meta("role").is_some());
    /// ```
    pub fn with_metadata(&self, updates: HashMap<String, Value>) -> Self {
        let mut meta = self.metadata.clone();
        meta.extend(updates);
        Self {
            name: self.name.clone(),
            content: self.content.clone(),
            metadata: meta,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn empty_name_is_invalid() {
        assert!(Block::new("", "content", HashMap::new()).is_err());
    }

    #[test]
    fn type_check() {
        let b = Block::new("system", "hi", HashMap::new()).unwrap();
        assert!(b.is_type("system"));
        assert!(!b.is_type("user"));
    }

    #[test]
    fn with_content_is_copy() {
        let b = Block::new("user", "old", HashMap::new()).unwrap();
        let b2 = b.with_content("new");
        assert_eq!(b.content, "old");
        assert_eq!(b2.content, "new");
    }

    #[test]
    fn meta_lookup() {
        let meta = [("k".to_string(), json!("v"))].into();
        let b = Block::new("user", "hi", meta).unwrap();
        assert_eq!(b.meta("k"), Some(&json!("v")));
        assert_eq!(b.meta("missing"), None);
    }
}
